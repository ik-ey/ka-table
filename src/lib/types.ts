import { AttributeTableData, Column } from './models';

import { ITableProps } from './Components/Table/Table';

export interface NoData {
  text?: string;
  hideHeader?: boolean;
}

type AddParameters<T, I> = T extends (e: infer E) => void ? (
    (e: E, extendedEvent: AttributeTableData<I>) => void
) : T;
type WithExtraParameters<T, I> = {
  [P in keyof T ] : AddParameters<T[P], I>;
};
type ElementAttributes<T> = React.AllHTMLAttributes<HTMLElement>;

export type ChildAttributesItem<T> = WithExtraParameters<ElementAttributes<T>, T> & { ref?: any };
export type DispatchFunc = (action: any) => void;
export type OnDispatchFunc = (action: any, tableProps: ITableProps) => void;
export type ControlledPropsKeys = (keyof ITableProps)[];
export type Field = string;
export type FormatFunc = (props: { value: any, column: Column }) => any;
export type FilterFunc = (props: { column: Column }) => ((value: any, filterRowValue: any, rowData?: any) => boolean) | void;
export type SortFunc = (props: { column: Column }) =>  ((value1: any, value2: any) => 0 | 1 | -1) | void;
export type SearchFunc = (props: { searchText: string, rowData: any, column: Column }) => boolean;
export type ValidationFunc = (props: { value: any, rowData: any, column: Column }) => string | void;
