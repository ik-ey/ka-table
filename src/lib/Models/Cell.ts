export class Cell {
    public columnKey!: string;
    public rowKeyValue!: any;
}
