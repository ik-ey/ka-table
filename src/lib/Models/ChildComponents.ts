import {
  ICellEditorProps,
  ICellProps,
  ICellTextProps,
  IDataRowProps,
  IEmptyCellProps,
  IFilterRowEditorProps,
  IGroupPanelCellProps,
  IGroupPanelProps,
  IGroupRowProps,
  IGroupSummaryCellProps,
  IGroupSummaryRowProps,
  IHeadCellProps,
  IHeadCellResizeProps,
  IHeadRowProps,
  IHeaderFilterButtonProps,
  ILoadingProps,
  INoDataRowProps,
  IPagingIndexProps,
  IPagingProps,
  IPagingSizeProps,
  IPopupContentItemProps,
  IPopupContentProps,
  IRowProps,
  ISortIconProps,
  ISummaryCellProps,
  ISummaryRowProps,
  ITableBodyProps,
  ITableFootProps,
  ITableHeadProps,
} from '../props';

import { ChildComponent } from './ChildComponent';
import { ITableProps } from '../';

export class ChildComponents {
  public cell?: ChildComponent<ICellProps>;
  public cellEditor?: ChildComponent<ICellEditorProps>;
  public cellEditorInput?: ChildComponent<ICellEditorProps>;
  public cellText?: ChildComponent<ICellTextProps>;
  public dataAndDetailsRows?: ChildComponent<IRowProps>;
  public dataRow?: ChildComponent<IDataRowProps>;
  public detailsCell?: ChildComponent<IDataRowProps>;
  public detailsRow?: ChildComponent<IDataRowProps>;
  public emptyCell?: ChildComponent<IEmptyCellProps>;
  public filterRowCell?: ChildComponent<IFilterRowEditorProps>;
  public groupPanel?: ChildComponent<IGroupPanelProps>;
  public groupPanelCell?: ChildComponent<IGroupPanelCellProps>;
  public groupExpandButton?: ChildComponent<IGroupRowProps>;
  public groupCell?: ChildComponent<IGroupRowProps>;
  public groupRow?: ChildComponent<IGroupRowProps>;
  public groupSummaryRow?: ChildComponent<IGroupSummaryRowProps>;
  public groupSummaryCell?: ChildComponent<IGroupSummaryCellProps>;
  public headFilterButton?: ChildComponent<IHeaderFilterButtonProps>;
  public headCell?: ChildComponent<IHeadCellProps>;
  public headCellContent?: ChildComponent<IHeadCellProps>;
  public headCellResize?: ChildComponent<IHeadCellResizeProps>;
  public headRow?: ChildComponent<IHeadRowProps>;
  public loading?: ChildComponent<ILoadingProps>;
  public noDataRow?: ChildComponent<INoDataRowProps>;
  public noDataCell?: ChildComponent<INoDataRowProps>;
  public paging?: ChildComponent<IPagingProps>;
  public pagingIndex?: ChildComponent<IPagingIndexProps>;
  public pagingPages?: ChildComponent<IPagingProps>;
  public pagingSize?: ChildComponent<IPagingSizeProps>;
  public pagingSizes?: ChildComponent<IPagingProps>;
  public popupContent?: ChildComponent<IPopupContentProps>;
  public popupContentItem?: ChildComponent<IPopupContentItemProps>;
  public rootDiv?: ChildComponent<ITableProps>;
  public sortIcon?: ChildComponent<ISortIconProps>;
  public summaryCell?: ChildComponent<ISummaryCellProps>;
  public summaryRow?: ChildComponent<ISummaryRowProps>;
  public table?: ChildComponent<ITableProps>;
  public tableBody?: ChildComponent<ITableBodyProps>;
  public tableFoot?: ChildComponent<ITableFootProps>;
  public tableHead?: ChildComponent<ITableHeadProps>;
  public tableWrapper?: ChildComponent<ITableProps>;
}
