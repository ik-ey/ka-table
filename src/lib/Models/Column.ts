import { DataType, SortDirection } from '../enums';

import { Field } from '../types';
import { PopupPosition } from './PopupPosition';

export class Column<TData = any> {
    public colGroup?: React.ColHTMLAttributes<HTMLElement>;
    public dataType?: DataType;
    public field?: Field;
    public filter?: (value: any, filterValue: any, rowData?: any) => boolean;
    public filterRowOperator?: any;
    public filterRowValue?: any;
    public headerFilterListItems?: (props: { data?: TData[] }) => string[];
    public headerFilterPopupPosition?: PopupPosition;
    public headerFilterSearch?: (value: any, searchValue: any, rowData?: any) => boolean;
    public headerFilterSearchValue?: any;
    public headerFilterValues?: string[];
    public isDraggable?: boolean;
    public isEditable?: boolean;
    public isFilterable?: boolean;
    public isHeaderFilterPopupShown?: boolean;
    public isHeaderFilterSearchable?: boolean;
    public isResizable?: boolean;
    public isSortable?: boolean;
    public key!: string;
    public sortDirection?: SortDirection;
    public sortIndex?: number;
    public style?: React.CSSProperties;
    public title?: string;
    public visible?: boolean;
    public width?: number | string;
}
