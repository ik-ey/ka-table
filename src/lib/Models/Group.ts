
export class Group {
    public columnKey!: string;
    public enableSummary?: boolean;
}
