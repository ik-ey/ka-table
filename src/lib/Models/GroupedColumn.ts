
export class GroupedColumn {
    public key!: string;
    public title?: string;
    public columnsKeys!: string[];
}
