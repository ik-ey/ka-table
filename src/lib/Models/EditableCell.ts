import { Cell } from './Cell';

export class EditableCell extends Cell {
    public editorValue?: any;
    public validationMessage?: any;
}
