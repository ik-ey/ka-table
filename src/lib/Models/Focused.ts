import { Cell } from './Cell';

export class Focused {
    public cell?: Cell;
    public cellEditorInput?: Cell;
}
