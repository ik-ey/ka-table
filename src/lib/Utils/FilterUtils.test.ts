import { DataType, FilterOperatorName } from '../enums';
import {
  filterData,
  getDefaultOperatorForType,
  getRowEditableCells,
  predefinedFilterOperators,
  searchData,
} from './FilterUtils';

import { Column } from '../Models/Column';
import { SearchFunc } from '../types';

describe('FilterUtils', () => {
  it('getRowEditableCells should return required cells from table EditableCells', () => {
    const rowEditableCells = getRowEditableCells(10, [{
      columnKey: '10',
      rowKeyValue: 10,
    }, {
      columnKey: '10',
      rowKeyValue: 2,
    }]);
    expect(rowEditableCells).toMatchSnapshot();
  });

  it('getDefaultOperatorForType', () => {
    expect(getDefaultOperatorForType(DataType.Number)).toBe('=');
    expect(getDefaultOperatorForType(DataType.Object)).toBe('=');
    expect(getDefaultOperatorForType(DataType.String)).toBe('contains');
  });

  describe('filterData', () => {
    const data: any[] = [
      { id: 1, name: 'Mike Wazowski', score: 80, passed: true, date: new Date(Date.UTC(2021, 11, 20, 9)) },
      { id: 2, name: 'Billi Bob', score: 55, passed: false, date: new Date(Date.UTC(2021, 11, 20, 13)) },
      { id: 3, name: 'Tom Williams', score: 45, passed: false, date: new Date(Date.UTC(2021, 10, 20, 13)) },
      { id: 4, name: 'Kurt Cobain', score: 75, passed: true, date: new Date(Date.UTC(2021, 11, 19, 13)) },
      { id: 5, name: 'Marshall Bruce', score: 77, passed: true, date: null },
      { id: 6, name: 'Sunny Fox', score: 33, passed: false, date: new Date(Date.UTC(2021, 11, 17, 13)) },
    ];
    it('one item', () => {
      const columns = [{
        filterRowValue: 'Billi Bob',
        key: 'name',
      }];
      const result = filterData(data, columns);
      expect(result).toMatchSnapshot();
    });

    it('by date', () => {
      const columnsDate: Column[] = [
        {
          dataType: DataType.Date,
          filterRowValue: new Date(Date.UTC(2021, 11, 20, 19, 18, 12)),
          key: 'date',
        },
      ];
      const result = filterData(data, columnsDate);
      expect(result).toMatchSnapshot();
    });

    it('IsEmpty for date', () => {
      const columnsDate: Column[] = [
        {
          dataType: DataType.Date,
          filterRowOperator: FilterOperatorName.IsEmpty,
          key: 'date',
        },
      ];
      const result = filterData(data, columnsDate);
      expect(result).toMatchSnapshot();
    });

    it('IsNotEmpty for date', () => {
      const columnsDate: Column[] = [
        {
          dataType: DataType.Date,
          filterRowOperator: FilterOperatorName.IsNotEmpty,
          key: 'date',
        },
      ];
      const result = filterData(data, columnsDate);
      expect(result).toMatchSnapshot();
    });

    it('two item', () => {
      const columns = [{
        filterRowValue: 'Billi Bob',
        key: 'name',
      }, {
        dataType: DataType.Number,
        filterRowValue: 45,
        key: 'score',
      }];
      const result = filterData(data, columns);
      expect(result).toMatchSnapshot();
    });

    it('custom filter', () => {
      const columns = [{
        dataType: DataType.Number,
        filterRowValue: 45,
        key: 'score',
      }];
      const result = filterData(data, columns, ({ column }) => {
        if (column.key === 'score'){
          return (value, filterValue) => value !== filterValue;
        }
      });
      expect(result).toMatchSnapshot();
    });

    it('custom filter: column with data type = "date" should not change filterRowValue', () => {
      const columns = [{
        dataType: DataType.Date,
        filterRowValue: ['date 1'],
        key: 'date',
      }];
      const result = filterData(data, columns, ({ column }) => {
        if (column.key === 'date'){
          return (value, filterValue) => {
            return filterValue[0] === 'date 1' && value === data[0].date
          };
        }
      });
      expect(result).toMatchSnapshot();
    });

    it('should throw an error in case of unknown filterOperator', () => {
      const columns = [{
        filterRowOperator: 'unknownOperator',
        filterRowValue: 'Billi Bob',
        key: 'name',
      }];
      expect(() => filterData(data, columns)).toThrowError('\'unknownOperator\' has not found in predefinedFilterOperators array, available operators: =, >, <, >=, <=, contains');
    });
  });

  [{
    falsy: [1, 2],
    name: '=',
    truthy: [1, 1],
  }, {
    falsy: [1, 1],
    name: '>',
    truthy: [2, 1],
  }, {
    falsy: [2, 1],
    name: '<',
    truthy: [1, 2],
  }, {
    falsy: [1, 2],
    name: '>=',
    truthy: [1, 1],
  }, {
    falsy: [2, 1],
    name: '<=',
    truthy: [1, 1],
  }, {
    falsy: ['abs', 'ss'],
    name: 'contains',
    truthy: ['hello', 'hell'],
  }, {
    falsy: [0, false],
    name: 'contains',
    truthy: [0, 0],
  }].forEach((d) => {
    it(`predefinedFilterOperators operator ${d.name} truthy: ${d.truthy} falsy: ${d.falsy}`, () => {
      const operator = predefinedFilterOperators.find((o) => o.name === d.name);
      if (!operator) {
        throw new Error(`${d.name} was not found`);
      }
      expect(operator.compare(d.truthy[0], d.truthy[1])).toBeTruthy();
      expect(operator.compare(d.falsy[0], d.falsy[1])).toBeFalsy();
    });
  });

  [{
    name: '=',
    nullShouldPass: false,
    undefinedShouldPass: false,
    value: 1,
  }, {
    name: '>',
    nullShouldPass: false,
    undefinedShouldPass: false,
    value: 1,
  }, {
    name: '<',
    nullShouldPass: true,
    undefinedShouldPass: false,
    value: 1,
  }, {
    name: '>=',
    nullShouldPass: false,
    undefinedShouldPass: false,
    value: 1,
  }, {
    name: '<=',
    nullShouldPass: true,
    undefinedShouldPass: false,
    value: 1,
  }, {
    name: 'contains',
    nullShouldPass: false,
    undefinedShouldPass: false,
    value: 1,
  }].forEach((d) => {
    it(`predefinedFilterOperators check null and undefined for ${d.name} operator`, () => {
      const operator = predefinedFilterOperators.find((o) => o.name === d.name);
      if (!operator) {
        throw new Error(`${d.name} was not found`);
      }
      expect(operator.compare(null, 1)).toEqual(d.nullShouldPass);
      expect(operator.compare(undefined, 1)).toEqual(d.undefinedShouldPass);
    });
  });

  describe('searchData', () => {
    const search: SearchFunc = ({ searchText, rowData, column }) => {
      if (column.key === 'passed')
        return (searchText === 'false' && !rowData.passed) || (searchText === 'true' && rowData.passed);
    };
    const columns: Column[] = [
      { key: 'name', title: 'Name', dataType: DataType.String },
      { key: 'score', title: 'Score', dataType: DataType.Number },
      {
        dataType: DataType.Boolean,
        key: 'passed',
        title: 'Passed',
      },
    ];
    const data: any[] = [
      { id: 1, name: 'Mike Wazowski', score: 80, passed: true },
      { id: 2, name: 'Billi Bob', score: 55, passed: false },
      { id: 3, name: 'Tom Williams', score: 45, passed: false },
      { id: 4, name: 'Kurt Cobain', score: 75, passed: true },
      { id: 5, name: 'Marshall Bruce', score: 77, passed: true },
      { id: 6, name: 'Sunny Fox', score: 33, passed: false },
      { id: 7, name: 'Falsey False', score: 33, passed: false },
    ];
    it('by string', () => {
      const result = searchData(columns, data, 'Mike', search);
      expect(result).toMatchSnapshot();
    });

    it('by boolean', () => {
      const columnsBoolean: Column[] = [
        {
          dataType: DataType.Boolean,
          key: 'passed',
          title: 'Passed',
        },
      ];
      const result = searchData(columnsBoolean, data, 'tru');
      expect(result).toMatchSnapshot();
    });

    it('should add item only once', () => {
      const result = searchData(columns, data, 'false', search);
      expect(result).toMatchSnapshot();
    });

    it('should not find value by search handler', () => {
      const result = searchData(columns, data, 'tru', search);
      expect(result).toMatchSnapshot();
    });

    it('should find value by search handler', () => {
      const result = searchData(columns, data, 'true', search);
      expect(result).toMatchSnapshot();
    });
  });
});
