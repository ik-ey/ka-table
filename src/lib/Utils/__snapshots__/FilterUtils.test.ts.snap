// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterUtils filterData IsEmpty for date 1`] = `
Array [
  Object {
    "date": null,
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
]
`;

exports[`FilterUtils filterData IsNotEmpty for date 1`] = `
Array [
  Object {
    "date": 2021-12-20T09:00:00.000Z,
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "date": 2021-12-20T13:00:00.000Z,
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
  Object {
    "date": 2021-11-20T13:00:00.000Z,
    "id": 3,
    "name": "Tom Williams",
    "passed": false,
    "score": 45,
  },
  Object {
    "date": 2021-12-19T13:00:00.000Z,
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "date": 2021-12-17T13:00:00.000Z,
    "id": 6,
    "name": "Sunny Fox",
    "passed": false,
    "score": 33,
  },
]
`;

exports[`FilterUtils filterData by date 1`] = `
Array [
  Object {
    "date": 2021-12-20T09:00:00.000Z,
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "date": 2021-12-20T13:00:00.000Z,
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
]
`;

exports[`FilterUtils filterData custom filter 1`] = `
Array [
  Object {
    "date": 2021-12-20T09:00:00.000Z,
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "date": 2021-12-20T13:00:00.000Z,
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
  Object {
    "date": 2021-12-19T13:00:00.000Z,
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "date": null,
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
  Object {
    "date": 2021-12-17T13:00:00.000Z,
    "id": 6,
    "name": "Sunny Fox",
    "passed": false,
    "score": 33,
  },
]
`;

exports[`FilterUtils filterData custom filter: column with data type = "date" should not change filterRowValue 1`] = `
Array [
  Object {
    "date": 2021-12-20T09:00:00.000Z,
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
]
`;

exports[`FilterUtils filterData one item 1`] = `
Array [
  Object {
    "date": 2021-12-20T13:00:00.000Z,
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
]
`;

exports[`FilterUtils filterData two item 1`] = `Array []`;

exports[`FilterUtils getRowEditableCells should return required cells from table EditableCells 1`] = `
Array [
  Object {
    "columnKey": "10",
    "rowKeyValue": 10,
  },
]
`;

exports[`FilterUtils searchData by boolean 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
]
`;

exports[`FilterUtils searchData by string 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
]
`;

exports[`FilterUtils searchData should add item only once 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Billi Bob",
    "passed": false,
    "score": 55,
  },
  Object {
    "id": 3,
    "name": "Tom Williams",
    "passed": false,
    "score": 45,
  },
  Object {
    "id": 6,
    "name": "Sunny Fox",
    "passed": false,
    "score": 33,
  },
  Object {
    "id": 7,
    "name": "Falsey False",
    "passed": false,
    "score": 33,
  },
]
`;

exports[`FilterUtils searchData should find value by search handler 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Mike Wazowski",
    "passed": true,
    "score": 80,
  },
  Object {
    "id": 4,
    "name": "Kurt Cobain",
    "passed": true,
    "score": 75,
  },
  Object {
    "id": 5,
    "name": "Marshall Bruce",
    "passed": true,
    "score": 77,
  },
]
`;

exports[`FilterUtils searchData should not find value by search handler 1`] = `Array []`;
